pip/pip-419.md

```diff
@@ -0,0 +1,180 @@
+ messageBuilder1.property("messageType", "file");
+ messageBuilder1.property("messageSource", "QQ");
+ 
+  // message2
+ TypedMessageBuilder<byte[]> messageBuilder2 = producer.newMessage();
+ messageBuilder2.property("messageType", "audio");
+ messageBuilder2.property("messageSource", "wechat");
+```
+
+If a subscription only wants to consume *audio* messages from w*echat*, how should it be implemented? 
+
+We can implement message filtering logic on pulsar broker at [org.apache.pulsar.broker.service.AbstractBaseDispatcher.filterEntriesForConsumer](https://github.com/apache/pulsar/blob/965ef5c14c93ca896ef4c8f34520066285fcf047/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java#L108)
+
+The filtering engine can choose the following two solutions. **In general, I prefer the SQL-92 solution, which should have better performance. Groovy may have slightly worse performance due to its dynamic characteristics.**
+
+### Solution 1 SQL-92
+
+Message filtering using [SQL-92](https://en.wikipedia.org/wiki/SQL-92) syntax mentioned in https://github.com/apache/pulsar/pull/8544#issuecomment-1064619792
+
+We can write the following SQL expression to filter messages, the pseudo code is as follow:
+
+```java
+// The filter expression passed by the pulsar consumer client to the pulsar broker
+String sql = "messageSource='wechat' AND messageType='audio'"
+
+// The pulsar broker uses the expression passed by the client to filter the messages and returns the messages that meet the conditions to the client.
+List filtedMessages = messages.stream()
+                    .filter(message -> {
+                        return SQL92Expression.evaluate(message)
+                    })
+                    .collect(Collectors.toList());
+```
+
+### Solution 2 Groovy
+
+> [Groovy](https://groovy-lang.org/) is a JVM-based object-oriented programming language under Apache. It can be used for object-oriented programming and can also be used as a pure scripting language to call Groovy scripts in Java.
+> 
+
+We can write the following Groovy expression to filter messages, the pseudo code is as follows:
+
+```java
+// The filter expression passed by the pulsar consumer client to the pulsar broker
+String groovyExpression = "it.properties.messageType.equals('audio') && it.properties.messageSource.equals('wechat')"
+
+// The pulsar broker uses the expression passed by the client to filter the messages and returns the messages that meet the conditions to the client.
+GroovyShell shell = new GroovyShell();
+Script script = shell.parse("return " + groovyExpression);
+List filtedMessages = messages.stream()
+                    .filter(message -> {
+                        binding.setVariable("it", message);
+                        script.setBinding(binding);
+                        return (Boolean) script.run();
+                    })
+                    .collect(Collectors.toList());
```

Comment:
This would be terrible from security perspective. Something like [Google CEL](https://github.com/google/cel-java) or [Spring Expression Language (SpEL)](https://docs.spring.io/spring-framework/reference/core/expressions.html) would be a better choice.